### ISSUE TITLE
`DeriveAnyClass` causes "Internal Error: initTcWithGbl failed"

### ISSUE THREAD

**Issue by MathiasSven** [haskell/haskell-language-server#3942]
When `DeriveAnyClass` is enabled, attempting to "hover" (inspect) a `deriving A` where `A` is a class with at least one method will cause HLS to have an internal error. 

### Your environment

Which OS do you use?
`Linux 6.7.0, NixOS, 24.05 (Uakari), 24.05.20240108.317484b`

Which version of GHC do you use, and how did you install it?
`Version 9.4.8, installed via nixpkgs`

How is your project built (alternative: link to the project)?
`N/A`

Which LSP client (editor/plugin) do you use?
`VS Codium+vscode-haskell`

Which version of HLS do you use, and how did you install it?
`haskell-language-server version: 2.4.0.0 (GHC: 9.4.8), nixpkgs`

Have you configured HLS in any way (especially: a `hie.yaml` file)?
`No`

### Steps to reproduce

```haskell
{-# LANGUAGE Haskell98 #-}
{-# LANGUAGE DeriveAnyClass #-}

module Main where

class C a where
  foo :: a -> Int

newtype Foo = MkFoo Int
  deriving Num

newtype Bar a = MkBar a
  deriving C

main :: IO ()
main = return ()
```

### Expected behaviour

The code above should only result in a `missing-methods` warning on `deriving Num` and `deriving C`

### Actual behaviour

Hovering over `Num` gives you the issue
```
• No explicit implementation for
    ‘+’, ‘*’, ‘abs’, ‘signum’, ‘fromInteger’, and (either ‘negate’
```
And over `C`
```
• No explicit implementation for
    ‘foo’
```

This is expected, however, when hovering over either HLS also gives an error condition pop-up every time they are hovered:
```
Error condition, please check your setup and/or the 
issue tracker: class: Internal Error: initTcWithGbl failed
```

**Comment by soulomoon** [comment]
Worth digging out. Maybe somewhere we forget to handle the error in ghc.

**Comment by soulomoon** [comment]
I found something really wired in hls-class-plugin.
```
isClassNodeIdentifier :: IdentifierDetails a -> Bool
isClassNodeIdentifier ident = (isNothing . identType) ident && Use `Set.member` identInfo ident
```

**Comment by soulomoon** [comment]
There are many holes in this plugin.

1. bug in isClassNodeIdentifier that do not classify idenfitifer correctly. Could be solve by #4020 .
2. also there is another bug, we should not invoke such code action in deriving expression, which require us to back trace to its parents to see if it is derving expression. not fixed by #4020 .
